zio-http/shared/src/main/scala/zio/http/codec/HttpCodec.scala

```diff
@@ -2264,140 +2269,220 @@ object HttpCodec extends ContentCodecs with HeaderCodecs with MethodCodecs with
-
+      if (isOptionalSchema) {
+        throw new IllegalArgumentException("Query is already optional")
+      } else {
+        Annotated(Query(codec.optional, index), Metadata.Optional())
       }
 
+    def tag: AtomTag = AtomTag.Query
+
   }
 
-  private[http] object Query {
-    sealed trait QueryType[A]
-    object QueryType {
-      case class Primitive[A](name: String, codec: BinaryCodecWithSchema[A]) extends QueryType[A]
-      case class Collection[A](colSchema: Schema.Collection[_, _], elements: QueryType.Primitive[A], optional: Boolean)
-          extends QueryType[A] {
-        def toCollection(values: Chunk[Any]): A =
-          colSchema match {
-            case Schema.Sequence(_, fromChunk, _, _, _) =>
-              fromChunk.asInstanceOf[Chunk[Any] => Any](values).asInstanceOf[A]
-            case Schema.Set(_, _)                       =>
-              values.toSet.asInstanceOf[A]
-            case _                                      =>
-              throw new IllegalArgumentException(
-                s"Unsupported collection schema for query object field of type: $colSchema",
-              )
-          }
-      }
-      case class Record[A](recordSchema: Schema[A]) extends QueryType[A] {
-        private var namesAndCodecs: Chunk[(Schema.Field[_, _], BinaryCodecWithSchema[Any])]       = _
-        private[http] def fieldAndCodecs: Chunk[(Schema.Field[_, _], BinaryCodecWithSchema[Any])] =
-          if (namesAndCodecs == null) {
-            namesAndCodecs = recordSchema match {
-              case record: Schema.Record[A]                =>
-                record.fields.map { field =>
-                  validateSchema(field.name, field.schema)
-                  val codec = binaryCodecForField(field.annotations.foldLeft(field.schema)(_ annotate _))
-                  (unlazy(field.asInstanceOf[Schema.Field[Any, Any]]), codec)
-                }
-              case s if s.isInstanceOf[Schema.Optional[_]] =>
-                val record = s.asInstanceOf[Schema.Optional[A]].schema.asInstanceOf[Schema.Record[A]]
-                record.fields.map { field =>
-                  validateSchema(field.name, field.annotations.foldLeft(field.schema)(_ annotate _))
-                  val codec = binaryCodecForField(field.schema)
-                  (field, codec)
-                }
-              case s => throw new IllegalArgumentException(s"Unsupported schema for query object field of type: $s")
-            }
-            namesAndCodecs
-          } else {
-            namesAndCodecs
-          }
+  object Query {
+    def apply[A](name: String, schema: Schema[A]): Query[A, A] = Query(SchemaCodec(Some(name), schema))
+    def apply[A](schema: Schema[A]): Query[A, A]               = Query(SchemaCodec(None, schema))
+  }
+
+  final case class SchemaCodec[A](name: Option[String], schema: Schema[A], kebabCase: Boolean = false) {
```

Comment:
Shouldn't this be made as private as possible? Including structure and methods? If it's really intended to be user-land, we'll want to review it more carefully and move it to the top level.